### PR TITLE
add wildcard support in xpack pipeline id

### PR DESCRIPTION
**Pull request by kaisecheng**
## What does this PR do?
allow `xpack.management.pipeline.id` to accept wildcard `*` setting to add/ update/ delete pipelines without restarting Logstash
add validation on `xpack.management.pipeline.id` to align with Kibana `pipeline ID must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers`
add warning msg if wildcard is used with elasticsearch version < 7.10

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## How to test this PR locally
** will need elasticsearch 7.10+, at this moment elastic cloud is 7.9, so unable to test in cloud**
- Logstash enable xpack 
- in logstash.yml `xpack.management.pipeline.id: ["*"]`
- start a local elasticsearch
- add a new pipeline to Elasticsearch
```
curl -X PUT -v "localhost:9200/_logstash/pipeline/heartbeat"  -H 'Content-Type: application/json' -d '{
    "pipeline": "input { heartbeat { } } output { stdout {} }",
    "last_modified": "2020-10-01T15:42:30.229Z",
    "pipeline_metadata": { "version": "4"},
    "username": "log.stash",
    "pipeline_settings": {"pipeline.batch.delay": "50"}
}'
```
- Logstash should pick up the new pipeline and start processing it

## Related issues
https://github.com/elastic/logstash/issues/10558

### CODE REVIEW

**Review by yaauie — COMMENTED** [review]
This looks to work as specified, but more on that below.

I'm wary of implementing the wildcarding on the Logstash side, and think it would be easier to maintain on the Elasticsearch side, querying Elasticsearch with the provided wildcards to get a list of pipelines instead of performing the filtering after fetching all pipelines. If we _do_ end up doing the client-side filtering, I would prefer [`File::fnmatch?`](https://ruby-doc.org/core-2.5.0/File.html#method-c-fnmatch), which handles traditional globbing and is safer with arbitrary user input than `Regexp::new`.

***

I would like to take a step back and consider how this feature (both as-specified and as-built) is likely to be used, and whether that provides value directly to our users.

I believe that the ultimate goal is for our users to be able to allocate specific pipelines to a specific already-running instance of Logstash, and to be able to push updates to those pipeline's definitions without restarting the Logstash process or affecting other unrelated pipelines. The current implementation fails to achieve this end because pipeline ids must be specified explicitly before process launch.

The wildcarding feature _kind-of_ improves upon this, but I don't believe it does enough to make it easy to reason about which process is running what from within Pipeline Management.

In order to achieve this end with the wildcarding feature, the user would need to start up each Logstash with a wildcard pattern that is distinct to that Logstash instance (e.g., `logstash-prod-alpha-*`) and essentially namespace their pipeline ids by which logstash instance they plan to run them on (e.g., `logstash-prod-alpha-syslog`). This largely precludes the ability for a single pipeline to run on multiple instances without duplication (and introduces possible drift between cloned definitions), and makes "moving" a pipeline from one instance to another an awkward renaming dance.

All of that said, this feature (both as-designed and as-implemented) does not seem to preclude an implementation where a Logstash instance _queries_ for which pipelines it should run. It is certainly valuable, but I believe we should also document expected usage patterns extensively and file issues/enhancement-requests to describe a more-direct solution.

File: x-pack/lib/config_management/elasticsearch_source.rb [hunk 1]
```diff
@@ -202,18 +203,47 @@ class SystemIndicesFetcher
...
+        wildcard_patterns, fix_pids = pipeline_ids.partition { |pattern| pattern.include?("*")}
+
+        fix_id_pipelines = fix_pids.map { |id|
+          response.has_key?(id) ? {id => response[id]}: {}
+        }.reduce({}, :merge)
+        fix_id_pipelines.keys.map { |id| response.delete(id)}
+
+        wildcard_matched_patterns = Set.new
+        wildcard_pattern_regex = wildcard_patterns.map { |pattern|
+          {pattern => Regexp.new("^" + pattern.gsub("*", ".*") + "$")}
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply x-pack/lib/config_management/elasticsearch_source.rb:233]: The pattern matching is base on the description from Kibana ` pipeline ID must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers`. I think you raise a good question of what if users don't follow the rule and create a pipeline directly through elasticsearch

We no longer be able to access Elasticsearch system indices directly in version 8. Instead, System indices API provides an endpoint to get a single pipeline and all pipelines without paging. Therefore wildcard query is not feasible anymore.

[the remark this answers is not in the document]
Comment (yaauie) [reply x-pack/lib/config_management/elasticsearch_source.rb]: Bummer, but that makes sense. We'll need to work with the APIs we have access to.

Let's validate the input before we pass it directly into whatever matcher we end up using.

The Setting is ArrayCoercible, so we can add a validator proc:

~~~ diff
diff --git a/x-pack/lib/config_management/extension.rb b/x-pack/lib/config_management/extension.rb
index 06d849439..f63d77b38 100644
--- a/x-pack/lib/config_management/extension.rb
+++ b/x-pack/lib/config_management/extension.rb
@@ -12,6 +12,9 @@ module LogStash
     class Extension < LogStash::UniversalPlugin
       include LogStash::Util::Loggable
 
+      # pipeline ID must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers
+      PIPELINE_ID_PATTERN = %r{\A[a-z_][a-z_\-0-9]]*\Z}i
+
       def register_hooks(hooks)
         require "logstash/runner"
         hooks.register_hooks(LogStash::Runner, Hooks.new)
@@ -23,7 +26,11 @@ module LogStash
 
         settings.register(LogStash::Setting::Boolean.new("xpack.management.enabled", false))
         settings.register(LogStash::Setting::TimeValue.new("xpack.management.logstash.poll_interval", "5s"))
-        settings.register(LogStash::Setting::ArrayCoercible.new("xpack.management.pipeline.id", String, ["main"]))
+        settings.register(LogStash::Setting::ArrayCoercible.new("xpack.management.pipeline.id", String, ["main"])) do |input_list|
+          invalid = input_list.reject { |entry| PIPELINE_ID_PATTERN =~ entry }
+          logger.warn("Pipeline IDs must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers.", :invalid_pipeline_ids => invalid) unless invalid.empty?
+          invalid.empty?
+        end
         settings.register(LogStash::Setting::NullableString.new("xpack.management.elasticsearch.username", "logstash_system"))
         settings.register(LogStash::Setting::NullableString.new("xpack.management.elasticsearch.password"))
         settings.register(LogStash::Setting::ArrayCoercible.new("xpack.management.elasticsearch.hosts", String, [ "https://localhost:9200" ] ))
~~~

Adding asterisks to this validation then becomes:

~~~ diff
diff --git a/x-pack/lib/config_management/extension.rb b/x-pack/lib/config_management/extension.rb
index f63d77b38..62a7f76b5 100644
--- a/x-pack/lib/config_management/extension.rb
+++ b/x-pack/lib/config_management/extension.rb
@@ -13,7 +13,8 @@ module LogStash
       include LogStash::Util::Loggable
 
       # pipeline ID must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers
-      PIPELINE_ID_PATTERN = %r{\A[a-z_][a-z_\-0-9]]*\Z}i
+      # wildcard character `*` is also acceptable and follows globbing rules
+      PIPELINE_ID_PATTERN = %r{\A[a-z_*][a-z_\-0-9*]]*\Z}i
 
       def register_hooks(hooks)
         require "logstash/runner"
@@ -28,7 +29,7 @@ module LogStash
         settings.register(LogStash::Setting::TimeValue.new("xpack.management.logstash.poll_interval", "5s"))
         settings.register(LogStash::Setting::ArrayCoercible.new("xpack.management.pipeline.id", String, ["main"])) do |input_list|
           invalid = input_list.reject { |entry| PIPELINE_ID_PATTERN =~ entry }
-          logger.warn("Pipeline IDs must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers.", :invalid_pipeline_ids => invalid) unless invalid.empty?
+          logger.warn("Pipeline IDs must begin with a letter or underscore and contain only letters, underscores, dashes, and numbers. The asterisk wildcard `*` can also be used.", :invalid_pipeline_ids => invalid) unless invalid.empty?
           invalid.empty?
         end
         settings.register(LogStash::Setting::NullableString.new("xpack.management.elasticsearch.username", "logstash_system"))
~~~

**Comment by kaisecheng** [comment]
This PR addresses to solve the need of restarting Logstash when adding a new pipeline. There are some suggestions for the future during the spec review. 

1. Enable users to config `xpack.management.pipeline.id` in Kibana
2. Be able to manage pipelines at the instance level in Kibana instead of all instances having the same configuration
3. Support delete and recreate case. Logstash unable to pick up the pipeline again if pipeline id and config string remain the same
4. Support pipeline versioning. Allow Logstash to rollback to the old version in case the new pipeline has an error.

As far as I know, number 2 is one of the most popular requests from users, and it is your concern as well. We can discuss again with @jsvd and @roaksoax if we can include it in this issue

File: x-pack/lib/config_management/elasticsearch_source.rb [hunk 2]
```diff
@@ -188,12 +183,18 @@ def client
     end
 
     module Fetcher
-      def config_exist?(pipeline_id)
-        @response.has_key?(pipeline_id)
+      include LogStash::Util::Loggable
+
+      def get_pipelines_key
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply x-pack/lib/config_management/elasticsearch_source.rb]: updated

File: x-pack/lib/config_management/elasticsearch_source.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaisecheng) [reply x-pack/lib/config_management/elasticsearch_source.rb]: I changed pattern matching to glob. Thanks for suggesting it.

[the remark this answers is not in the document]
Comment (kaisecheng) [reply x-pack/lib/config_management/elasticsearch_source.rb]: I added the validation as suggested. One concern is if users have a pipeline id with an invalid character like `.`, they will fail the checking and unable to start Logstash

File: x-pack/lib/config_management/elasticsearch_source.rb [hunk 3]
```diff
@@ -202,18 +203,44 @@ class SystemIndicesFetcher
       SYSTEM_INDICES_API_PATH = "_logstash/pipeline"
 
       def fetch_config(pipeline_ids, client)
-        path_ids = pipeline_ids.join(",")
-        response = client.get("#{SYSTEM_INDICES_API_PATH}/#{path_ids}")
+        response = client.get("#{SYSTEM_INDICES_API_PATH}/")
```
Comment (yaauie) [inline x-pack/lib/config_management/elasticsearch_source.rb:206]: Is this the API that was introduced in Elasticsearch 7.10? If so, I think we need to conditionally query _with_ the pipeline ids if no wildcard is present. That way Logstash continues to work with older Elasticsearches unless explicitly configured to use this feature.

```suggestion
        has_wildcard = pipeline_ids.any? { |id| id.include?("*") }
        if has_wildcard
          logger.warn("wildcard '*' in xpack.management.pipeline.id is not supported in Elasticsearch version < 7.10. If you are using Elasticsearch 7.10+ this message can safely be ignored.")
          response = client.get("#{SYSTEM_INDICES_API_PATH}/")
        else
          path_ids = pipeline_ids.join(",")
          response = client.get("#{SYSTEM_INDICES_API_PATH}/#{path_ids}")
        end
```

**Review by yaauie — COMMENTED** [review]
I think we're on the right track, but may have missed a backward-compatibility issue with older Elasticsearch instances in the 7.x series, because we query an API that was introduced in Elasticsearch 7.10.

File: x-pack/lib/config_management/elasticsearch_source.rb (hunk 3 above)
Comment (kaisecheng) [reply x-pack/lib/config_management/elasticsearch_source.rb:206]: logstash does not call elasticsearch < 7.10 with `SystemIndicesFetcher`. 
backward compatibility with `LegacyHiddenIndicesFetcher` is here https://github.com/elastic/logstash/pull/12279/files#diff-26a8304237e5355c83934e069ef00ae959860a114476aa61e019f913131d8b13R64

Comment (yaauie) [reply x-pack/lib/config_management/elasticsearch_source.rb:206]: Ah HA. That is the bit I was missing.

**Review by yaauie — APPROVED** [review]
LGTM 👍

**Pull request merged**